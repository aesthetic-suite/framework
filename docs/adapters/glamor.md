# Glamor

Provides [Glamor](https://github.com/threepointone/glamor) support.

```javascript
import Aesthetic from 'aesthetic';
import GlamorAdapter from 'aesthetic-adapter-glamor';

const aesthetic = new Aesthetic(new GlamorAdapter());
```

## Unified Syntax

To make use of the unified syntax, simply import the adapter from the unified path.

```javascript
import GlamorAdapter from 'aesthetic-adapter-glamor/unified';
```

## Plugins

[Glamor plugins](https://github.com/threepointone/glamor/blob/master/docs/plugins.md) can be added
before the adapter is instantiated.

```javascript
import Aesthetic from 'aesthetic';
import GlamorAdapter from 'aesthetic-adapter-glamor';
import { plugins } from 'glamor';

plugins.add(plugin); // ...

const aesthetic = new Aesthetic(new GlamorAdapter());
```
