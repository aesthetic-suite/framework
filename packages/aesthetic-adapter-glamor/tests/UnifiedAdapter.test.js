/* eslint-disable sort-keys */

import { speedy, flush } from 'glamor';
import UnifiedGlamorAdapter from '../src/UnifiedAdapter';
import {
  SYNTAX_UNIFIED_FULL,
  SYNTAX_FALLBACK,
  SYNTAX_AT_RULES,
  SYNTAX_PSEUDO,
  SYNTAX_FONT_FACE,
  SYNTAX_KEYFRAMES,
  SYNTAX_MEDIA_QUERY,
  SYNTAX_SUPPORTS,
} from '../../../tests/mocks';

describe('aesthetic-adapter-glamor/UnifiedAdapter', () => {
  let instance;

  beforeEach(() => {
    flush();
    speedy(true);

    instance = new UnifiedGlamorAdapter();
  });

  it('transforms style declarations into class names', () => {
    expect(instance.transform('component', SYNTAX_UNIFIED_FULL)).toEqual({
      button: 'css-bu86dt',
    });
  });

  it('converts unified syntax to native syntax', () => {
    expect(instance.convert(SYNTAX_UNIFIED_FULL)).toEqual({
      button: {
        margin: 0,
        padding: '6px 12px',
        border: '1px solid #2e6da4',
        borderRadius: 4,
        display: 'inline-block',
        cursor: 'pointer',
        fontFamily: 'Roboto',
        fontWeight: 'normal',
        lineHeight: 'normal',
        whiteSpace: 'nowrap',
        textDecoration: 'none',
        textAlign: 'center',
        backgroundColor: '#337ab7',
        verticalAlign: 'middle',
        color: ['#fff', 'rgba(0, 0, 0, 0)'],
        animationName: 'fade_1q3syk4',
        animationDuration: '.3s',
        ':hover': {
          backgroundColor: '#286090',
          borderColor: '#204d74',
        },
        '::before': {
          content: '"â˜…"',
          display: 'inline-block',
          verticalAlign: 'middle',
          marginRight: 5,
        },
        '@media (max-width: 600px)': {
          padding: '4px 8px',
        },
        '@supports (display: flex)': {
          display: 'flex',
        },
      },
    });
  });

  it('allows standard at-rules', () => {
    expect(instance.convert(SYNTAX_AT_RULES)).toEqual(SYNTAX_AT_RULES);
  });

  it('supports pseudos', () => {
    expect(instance.convert(SYNTAX_PSEUDO)).toEqual(SYNTAX_PSEUDO);
  });

  it('supports fallbacks', () => {
    expect(instance.convert(SYNTAX_FALLBACK)).toEqual({
      fallback: {
        background: ['red', 'linear-gradient(...)'],
        display: ['box', 'flex-box', 'flex'],
      },
    });
  });

  it('supports font faces', () => {
    expect(instance.convert(SYNTAX_FONT_FACE)).toEqual({
      font: {
        fontFamily: 'Roboto',
        fontSize: 20,
      },
    });
  });

  it('supports animations', () => {
    expect(instance.convert(SYNTAX_KEYFRAMES)).toEqual({
      animation: {
        animationName: 'fade_1q3syk4',
        animationDuration: '3s, 1200ms',
        animationIterationCount: 'infinite',
      },
    });
  });

  it('supports media queries', () => {
    expect(instance.convert(SYNTAX_MEDIA_QUERY)).toEqual({
      media: {
        color: 'red',
        '@media (max-width: 1000px)': {
          color: 'green',
        },
        '@media (min-width: 300px)': {
          color: 'blue',
        },
      },
    });
  });

  it('supports supports', () => {
    expect(instance.convert(SYNTAX_SUPPORTS)).toEqual({
      sup: {
        display: 'block',
        '@supports (display: flex)': {
          display: 'flex',
        },
        '@supports not (display: flex)': {
          float: 'left',
        },
      },
    });
  });
});
