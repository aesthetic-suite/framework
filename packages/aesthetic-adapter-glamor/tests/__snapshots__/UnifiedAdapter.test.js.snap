// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles @fallbacks 1`] = `".css-1806hfp,[data-css-1806hfp] {background: red; display: flex-box;}"`;

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles @font-face 1`] = `"@font-face {font-family: Roboto; font-style: normal; font-weight: normal; src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}.css-1x6s9dk,[data-css-1x6s9dk] {font-family: Roboto; font-size: 20px;}"`;

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles @keyframes 1`] = `"@-webkit-keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }@-moz-keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }@-o-keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }@keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }.css-s8bawe,[data-css-s8bawe] {animation-name: fade_1q3syk4; animation-duration: 3s, 1200ms; animation-iteration-count: infinite; -webkit-animation-name: fade_1q3syk4; -webkit-animation-duration: 3s, 1200ms; -webkit-animation-iteration-count: infinite;}"`;

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles @media 1`] = `".css-rr71yy,[data-css-rr71yy] {color: red;}@media (min-width: 300px) {.css-rr71yy,[data-css-rr71yy] {color: blue;}}@media (max-width: 1000px) {.css-rr71yy,[data-css-rr71yy] {color: green;}}"`;

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles @supports 1`] = `".css-1sp1mbh,[data-css-1sp1mbh] {display: block;}"`;

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles descendant selectors 1`] = `".css-iq5gps,[data-css-iq5gps] {margin: 0; padding: 0;}.css-iq5gps> li,[data-css-iq5gps]> li {list-style: bullet;}"`;

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles properties 1`] = `".css-115cnno,[data-css-115cnno] {color: black; display: inline; margin: 10px;}"`;

exports[`aesthetic-adapter-glamor/UnifiedAdapter handles pseudos 1`] = `".css-1g7aevf,[data-css-1g7aevf] {position: fixed;}.css-1g7aevf:hover,[data-css-1g7aevf]:hover {position: static;}.css-1g7aevf::before,[data-css-1g7aevf]::before {position: absolute;}"`;
